### PR TITLE
mds: fix `dump dir` indicate dir is not cached

### PR DESCRIPTION
**Pull request by joscollin**
Fixes: https://tracker.ceph.com/issues/63093

### CODE REVIEW

File: src/mds/MDSRank.cc [hunk 1]
```diff
@@ -3387,8 +3387,14 @@ void MDSRank::command_dump_dir(Formatter *f, const cmdmap_t &cmdmap, std::ostrea
   in->dirfragtree.get_leaves_under(frag_t(), leaves);
   for (const auto& leaf : leaves) {
     CDir *dir = in->get_dirfrag(leaf);
-    if (dir)
+    if (dir) {
       mdcache->dump_dir(f, dir, dentry_dump);
+    }
+    else {
```
Comment (batrick) [inline src/mds/MDSRank.cc:3393]:
```suggestion
    } else {
```

style

File: src/mds/MDSRank.cc [hunk 2]
```diff
@@ -3387,8 +3387,14 @@ void MDSRank::command_dump_dir(Formatter *f, const cmdmap_t &cmdmap, std::ostrea
...
+      f->open_object_section("frag");
+      f->dump_string("status", "dirfrag not in cache");
```
Comment (batrick) [inline src/mds/MDSRank.cc:3395]: How do you know which dirfrag is not in cache?

**Review by batrick — CHANGES_REQUESTED** [review]
Please include an example output in the commit message.

File: src/mds/MDSRank.cc [hunk 3]
```diff
@@ -3393,8 +3393,16 @@ void MDSRank::command_dump_dir(Formatter *f, const cmdmap_t &cmdmap, std::ostrea
...
   for (const auto& leaf : leaves) {
     CDir *dir = in->get_dirfrag(leaf);
-    if (dir)
+    if (dir) {
       mdcache->dump_dir(f, dir, dentry_dump);
+    } else {
+      f->open_object_section("frag");
+      CachedStackStringStream css;
+      *css << std::hex << leaf.value() << "/" << std::dec << leaf.bits();
+      f->dump_string("value/bits", css->strv());
```
Comment (batrick) [inline src/mds/MDSRank.cc:3402]:
```suggestion
      f->dump_stream("frag") << leaf;
```

**Review by batrick — CHANGES_REQUESTED** [review]

**Comment by joscollin** [comment]
jenkins test make check

**Review by vshankar — APPROVED** [review]
LGTM.

**Review by dparmar18 — APPROVED** [review]

**Comment by vshankar** [comment]
https://pulpito.ceph.com/?branch=wip-vshankar-testing-20240226.124304

**Comment by vshankar** [comment]
https://pulpito.ceph.com/vshankar-2024-02-27_04:05:06-fs-wip-vshankar-testing-20240226.124304-testing-default-smithi/

**Comment by vshankar** [comment]
Reruning tests with https://github.com/ceph/ceph/pull/55909 and https://github.com/ceph/ceph/pull/55908 included. See - https://pulpito.ceph.com/vshankar-2024-03-04_08:26:39-fs-wip-vshankar-testing-20240304.042522-testing-default-smithi/

**Review by vshankar — APPROVED** [review]
https://tracker.ceph.com/projects/cephfs/wiki/Main#5th-March-2024

**Pull request merged**